### PR TITLE
fix(sqllab): async query broken due to #21320

### PR DESCRIPTION
**Pull request by justinpark**
### SUMMARY
This commit fixes the async query broken issue on #21656

In #21320, it updated the following code 

```diff
---624      runQuery={this.runQuery}
+++540      runQuery={startQuery}
```
which skips the following `startQuery` with empty args operation.
```
/// 422
  runQuery() {
    if (this.props.database) {
      this.startQuery();
    }
  }
```
This different payload broke the async query request.
This commit reverts back this origin logic.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

- Before

Screen Shot 2022-09-30 at 2 00 34 PM

- After

Screen Shot 2022-09-30 at 2 00 59 PM

### TESTING INSTRUCTIONS
enable database option to async query execution
run the query

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/SqlLab/components/SqlEditor/index.jsx [hunk 1]
```diff
@@ -229,13 +229,20 @@ const SqlEditor = ({
...
 
-  useState(() => {
+  const runQuery = () => {
+    if (database) {
+      startQuery();
+    }
+  };
+
+  // hack the useMemo hook to imitate componentWillMount
+  useMemo(() => {
```
Comment (eric-briscoe) [inline superset-frontend/src/SqlLab/components/SqlEditor/index.jsx:239]: I think the  `useEffect` hook be more appropriate here and not be considered a hack, but recommended practice from React.

When using a `
`
useEffect() => {
}, [])
`
With empty array it tells the functional component to only run the effect one time.

Notes section of: https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects

"If you want to run an effect and clean it up only once (on mount and unmount), you can pass an empty array ([]) as a second argument. This tells React that your effect doesn’t depend on any values from props or state, so it never needs to re-run. This isn’t handled as a special case — it follows directly from how the dependencies array always works." 

Basically same exact code just swapping out the `useMemo` with `useEffect` to align with React's best practices and remove the hack comment

**Review by eric-briscoe — CHANGES_REQUESTED** [review]

File: superset-frontend/src/SqlLab/components/SqlEditor/index.jsx (hunk 1 above)
Comment (EugeneTorap) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.jsx:239]: @eric-briscoe am I correct
useMemo(() => {}, []); is componentWillMount event.
useEffect(() => {}, []); is componentDidMount

**Review by zhaoyongjie — APPROVED** [review]
work as expected, tested in my local env. Thanks for the quick fix!

### The master branch
async query - before

### The current PR

async query - after

**Comment by EugeneTorap** [comment]
@justinpark Can we try to cover it with RTL test?

File: superset-frontend/src/SqlLab/components/SqlEditor/index.jsx (hunk 1 above)
Comment (eric-briscoe) [reply superset-frontend/src/SqlLab/components/SqlEditor/index.jsx:239]: @EugeneTorap `useEffect` is intended to replace all of the Class lifecycle methods so with functional components React is saying you no longer would try to differentiate between componentDidMount, componentDidUpdate, and componentWillUnmount, just use `useEffect`.

`Tip If you’re familiar with React class lifecycle methods, you can think of useEffect Hook as componentDidMount, componentDidUpdate, and componentWillUnmount combined.` https://reactjs.org/docs/hooks-effect.html 

useMemo is for a very specific purpose and even React recommends not using it until you have a working component and start experience performance issues. useMemo runs during render and is not intended to produce side effects.

`Remember that the function passed to useMemo runs during rendering. Don’t do anything there that you wouldn’t normally do while rendering. For example, side effects belong in useEffect, not useMemo.` https://reactjs.org/docs/hooks-reference.html#usememo

**Review by villebro — APPROVED** [review]
